extensions/lem-dashboard/lem-dashboard.lisp

```diff
@@ -0,0 +1,128 @@
+         (spaces (make-string padding :initial-element #\Space)))
+    (concatenate 'string spaces str)))
+
+(defclass dashboard-item ()
+  ((item-attribute 
+    :initarg :item-attribute 
+    :accessor item-attribute 
+    :initform 'document-text-attribute
+    :documentation "Attribute to use when drawing this item.")
+   (top-margin
+    :initarg :top-margin 
+    :accessor top-margin
+    :initform 0
+    :documentation "The amount of vertical space (lines) to apply before the item.")
+   (bottom-margin
+    :initarg :bottom-margin 
+    :accessor bottom-margin
+    :initform 1
+    :documentation "The amount of vertical space (lines) to apply after the item.")
+   (action 
+    :initarg :action 
+    :accessor action 
+    :initform nil
+    :documentation "Function to execute when <return> is pressed over this item."))
+  (:documentation "Base class for all dashboard items."))
+
+(defgeneric draw-dashboard-item (item point)
+  (:documentation "Called to draw the dashboard item.")
+  (:method :before ((item dashboard-item) point)
+    (dotimes (i (top-margin item))
+      (insert-character point #\Newline)))
+  (:method :after ((item dashboard-item) point)
+    (dotimes (i (bottom-margin item))
+      (insert-character point #\Newline))))
+
+(define-command dashboard-open-selected-item () ()
+  "Execute action on selected dashboard item."
+  (let* ((point (current-point))
+         (item (text-property-at point :dashboard-item)))
+    (when (and item (action item))
+      (funcall (action item)))))
+
+(defmethod draw-dashboard-item :around ((item dashboard-item) point)
+  "Inserts a :dashboard-item text property in between the starting and ending POINT, useful for tracking."
+  (let ((start (copy-point point :temporary)))
+    (call-next-method)
+    (let ((end (copy-point point :temporary)))
+      (put-text-property start end :dashboard-item item)
+      (delete-point start)
+      (delete-point end))))
+
+(define-major-mode dashboard-mode ()
+    (:name "Dashboard"
+     :keymap *dashboard-mode-keymap*))
+
+(defun create-dashboard-buffer ()
+  "Creates the dashboard buffer."
+  (or (get-buffer *dashboard-buffer-name*)
+      (make-buffer *dashboard-buffer-name*
+                   :enable-undo-p nil
+                   :read-only-p t)))
+
+(defun redraw-dashboard ()
+  "Redraws the dashboard, clearing and redrawing all content while attempting to maintain point position."
+  (let* ((buffer (create-dashboard-buffer))
+         (old-line (line-number-at-point (buffer-point buffer)))
+         (old-column (point-column (buffer-point buffer))))
+    (setf (buffer-read-only-p buffer) nil)
+    (erase-buffer buffer)
+    (let ((point (buffer-point buffer)))
+      (dolist (item *dashboard-layout*)
+        (draw-dashboard-item item point)))
+    (setf (buffer-read-only-p buffer) t)
+    (change-buffer-mode buffer 'dashboard-mode)
+    (move-to-line (buffer-point buffer) old-line)
+    (move-to-column (buffer-point buffer) old-column)))
```

Comment:
Was there a problem with save-excursion?

Reply:
It didn't work for me... I might be using it wrong. Here's the redraw function, works fine (keeps cursor position when resizing window)

```lisp
(defun redraw-dashboard ()
  (let* ((buffer (create-dashboard-buffer))
         (old-line (line-number-at-point (buffer-point buffer)))
         (old-column (point-column (buffer-point buffer))))
    (with-buffer-read-only buffer nil
      (erase-buffer buffer)
      (let ((point (buffer-point buffer)))
        (dolist (item *dashboard-layout*)
          (draw-dashboard-item item point)))
      (change-buffer-mode buffer 'dashboard-mode)
      (move-to-line (buffer-point buffer) old-line)
      (move-to-column (buffer-point buffer) old-column))))
```

Here's after changing it:

```lisp
(defun redraw-dashboard ()
  (let* ((buffer (create-dashboard-buffer)))
    (save-excursion
      (with-buffer-read-only buffer nil
        (erase-buffer buffer)
        (let ((point (buffer-point buffer)))
          (dolist (item *dashboard-layout*)
            (draw-dashboard-item item point)))
        (change-buffer-mode buffer 'dashboard-mode)))))
```

After making that change, if I resize the window it'll reset it to the top. I tried using it in a few other spots and couldn't get it to work.